### PR TITLE
Add `view` kwarg to `first` and `last`

### PR DESCRIPTION
**Pull request by Chandu-4444**
Please look at this, I've added `view` (defaults to false) kwarg to `first` and `last`. Also added tests to check for checking types of the results that were returned by the above functions. https://github.com/JuliaData/DataFrames.jl/issues/2845

### CODE REVIEW

File: test/dataframe.jl [hunk 1]
```diff
@@ -1105,6 +1105,11 @@ end
     @test last(df, 6) == DataFrame(A=5:10)
     @test last(df, 1) == DataFrame(A=10)
 
+    @test first(df, 6, true) isa SubDataFrame
```
Comment (bkamins) [inline test/dataframe.jl:1108]: als add tests that the returned data frame has correct contents when `view=true`.

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -486,11 +486,16 @@ Get the first row of `df` as a `DataFrameRow`.
...
-    first(df::AbstractDataFrame, n::Integer)
+    first(df::AbstractDataFrame, n::Integer, view::Bool=false)
 
 Get a data frame with the `n` first rows of `df`.
+
+If `view=false` a freshly allocated `DataFrame` is returned.
+If `view=true` then a `SubDataFrame` view into `df` is returned.
 """
-Base.first(df::AbstractDataFrame, n::Integer) = df[1:min(n, nrow(df)), :]
+function Base.first(df::AbstractDataFrame, n::Integer, view::Bool=false)
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:496]: I think that `view` should be a kwarg. @nalimilan - can you please say what you prefer before we make the change?

File: src/abstractdataframe/abstractdataframe.jl [hunk 3]
```diff
@@ -486,11 +486,16 @@ Get the first row of `df` as a `DataFrameRow`.
...
+    first(df::AbstractDataFrame, n::Integer, view::Bool=false)
 
 Get a data frame with the `n` first rows of `df`.
+
+If `view=false` a freshly allocated `DataFrame` is returned.
+If `view=true` then a `SubDataFrame` view into `df` is returned.
 """
-Base.first(df::AbstractDataFrame, n::Integer) = df[1:min(n, nrow(df)), :]
+function Base.first(df::AbstractDataFrame, n::Integer, view::Bool=false) 
+    return view ? Base.view(df, 1:min(n ,nrow(df)), :) : df[1:min(n, nrow(df)), :]
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:497]: better keep the `=` style as this is a single expression.

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -493,9 +493,7 @@ Get a data frame with the `n` first rows of `df`.
 If `view=false` a freshly allocated `DataFrame` is returned.
 If `view=true` then a `SubDataFrame` view into `df` is returned.
 """
-function Base.first(df::AbstractDataFrame, n::Integer, view::Bool=false) 
-    return view ? Base.view(df, 1:min(n ,nrow(df)), :) : df[1:min(n, nrow(df)), :]
-end
+Base.first(df::AbstractDataFrame, n::Integer, view::Bool=false) = view ? Base.view(df, 1:min(n ,nrow(df)), :) : df[1:min(n, nrow(df)), :]
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:496]:
```suggestion
Base.first(df::AbstractDataFrame, n::Integer, view::Bool=false) =
    view ? Base.view(df, 1:min(n ,nrow(df)), :) : df[1:min(n, nrow(df)), :]
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 5]
```diff
@@ -512,9 +510,7 @@ Get a data frame with the `n` last rows of `df`.
 If `view=false` a freshly allocated `DataFrame` is returned.
 If `view=true` then a `SubDataFrame` view into `df` is returned.
 """
-function Base.last(df::AbstractDataFrame, n::Integer, view::Bool=false)
-    return view ? Base.view(df, max(1, nrow(df)-n+1):nrow(df), :) : df[max(1, nrow(df)-n+1):nrow(df), :]
-end
+Base.last(df::AbstractDataFrame, n::Integer, view::Bool=false) = view ? Base.view(df, max(1, nrow(df)-n+1):nrow(df), :) : df[max(1, nrow(df)-n+1):nrow(df), :]
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:513]:
```suggestion
Base.last(df::AbstractDataFrame, n::Integer, view::Bool=false) =
    view ? Base.view(df, max(1, nrow(df)-n+1):nrow(df), :) : df[max(1, nrow(df)-n+1):nrow(df), :]
```

**Comment by nalimilan** [comment]
It would be good to have an `@inferred` test if that's doable.

**Comment by bkamins** [comment]
> It would be good to have an @inferred test if that's doable.

My experience is that it will be problematic, but we can check :).

File: src/abstractdataframe/abstractdataframe.jl (hunk 2 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl]: OK - so given the 👍 - please change it to kwarg.

**Comment by Chandu-4444** [comment]
I didn't clearly understand the working of `@inferred`. Is it possible to point a few resources, I tried this https://docs.julialang.org/en/v1/stdlib/Test/#Test.@inferred, but still I wasn't very obvious about this.

**Comment by bkamins** [comment]
`@inferred` checks if Julia compiler is able to determine the return value of the function.

In our case here the question is if compiler is able to determine that when you do not pass `view` kwarg the returned value will have `DataFrame` type.

So just check if `@inferred first(df, 6)` errors or not. If it errors it is not the end of the world. For example with `filter` we have:
```
julia> df = DataFrame(x=1:3)
3×1 DataFrame
 Row │ x
     │ Int64
─────┼───────
   1 │     1
   2 │     2
   3 │     3

julia> @inferred filter(x -> true, df)
3×1 DataFrame
 Row │ x
     │ Int64
─────┼───────
   1 │     1
   2 │     2
   3 │     3

julia> @inferred filter(x -> true, df, view=false)
ERROR: return type DataFrame does not match inferred return type Union{DataFrame, SubDataFrame{DataFrame, DataFrames.Index, T} where T<:AbstractVector{Int64}}
```
and both results are expected and optimal. By default compiler knows that the returned value will have `DataFrame` type, however, when you pass `view` kwarg the return type is a union as at compile time the value of `view` kwarg cannot be determined.

File: test/dataframe.jl [hunk 6]
```diff
@@ -1104,6 +1104,13 @@ end
     @test first(df, 1) == DataFrame(A=1)
     @test last(df, 6) == DataFrame(A=5:10)
     @test last(df, 1) == DataFrame(A=10)
+    @test first(df, 6, true) == DataFrame(A=1:6)
```
Comment (bkamins) [inline test/dataframe.jl:1107]: fix tests; they should be `view=...`

**Comment by Chandu-4444** [comment]
> So just check if `@inferred first(df, 6)` errors or not.
 
Yes, that throws an error.

File: src/abstractdataframe/abstractdataframe.jl [hunk 7]
```diff
@@ -486,11 +486,15 @@ Get the first row of `df` as a `DataFrameRow`.
...
-    first(df::AbstractDataFrame, n::Integer)
+    first(df::AbstractDataFrame, n::Integer; view::Bool=false)
 
 Get a data frame with the `n` first rows of `df`.
+
+If `view=false` a freshly allocated `DataFrame` is returned.
+If `view=true` then a `SubDataFrame` view into `df` is returned.
 """
-Base.first(df::AbstractDataFrame, n::Integer) = df[1:min(n, nrow(df)), :]
+Base.first(df::AbstractDataFrame, n::Integer; view::Bool=false) =
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:496]:
```suggestion
@inline Base.first(df::AbstractDataFrame, n::Integer; view::Bool=false) = 
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 8]
```diff
@@ -500,11 +504,15 @@ Get the last row of `df` as a `DataFrameRow`.
...
-    last(df::AbstractDataFrame, n::Integer)
+    last(df::AbstractDataFrame, n::Integer; view::Bool=false)
 
 Get a data frame with the `n` last rows of `df`.
+
+If `view=false` a freshly allocated `DataFrame` is returned.
+If `view=true` then a `SubDataFrame` view into `df` is returned.
 """
-Base.last(df::AbstractDataFrame, n::Integer) = df[max(1, nrow(df)-n+1):nrow(df), :]
+Base.last(df::AbstractDataFrame, n::Integer; view::Bool=false) =
```
Comment (bkamins) [inline src/abstractdataframe/abstractdataframe.jl:514]:
```suggestion
@inline Base.last(df::AbstractDataFrame, n::Integer; view::Bool=false) = 
```

File: test/dataframe.jl [hunk 9]
```diff
@@ -1104,6 +1104,13 @@ end
     @test first(df, 1) == DataFrame(A=1)
     @test last(df, 6) == DataFrame(A=5:10)
     @test last(df, 1) == DataFrame(A=10)
+    @test first(df, 6, view=true) == DataFrame(A=1:6)
```
Comment (bkamins) [inline test/dataframe.jl:1107]:
```suggestion

    @inferred first(df, 6)
    @inferred last(df, 6)
    @inferred first(df)
    @inferred first(df)

    @test first(df, 6, view=true) == DataFrame(A=1:6)
```

File: test/dataframe.jl [hunk 10]
```diff
@@ -1105,6 +1105,19 @@ end
     @test last(df, 6) == DataFrame(A=5:10)
     @test last(df, 1) == DataFrame(A=10)
 
+    @inferred first(df, 6)
+    @inferred last(df, 6)
+    @inferred first(df)
+    @inferred first(df)
```
Comment (bkamins) [inline test/dataframe.jl:1111]:
```suggestion
    @inferred last(df)
```

**Comment by bkamins** [comment]
I have pushed changes that should make it `@inferred` hopefully.

**Comment by bkamins** [comment]
Yes - my fix works.

**Review by bkamins — APPROVED** [review]
Looks good (subject to wording of docstrings, which @nalimilan is better at checking than me)

File: NEWS.md [hunk 11]
```diff
@@ -94,6 +94,8 @@
   ([#2459](https://github.com/JuliaData/DataFrames.jl/pull/2459))
 * `sort!` now supports general `AbstractDataFrame`
   ([#2946](https://github.com/JuliaData/DataFrames.jl/pull/2946))
+* `filter` now supports `view` keyword argument
```
[the remark this answers is not in the document]
Comment (Chandu-4444) [reply NEWS.md:97]: Sure, somehow I forgot to update it this time.

**Review by nalimilan — APPROVED** [review]
Thanks!

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**